### ISSUE TITLE
Bug saving/loading proofs

### ISSUE THREAD

**Issue by wlcsm** [zxcalc/zxlive#207]
There is a bug where sometimes proofs are not saved or loaded correctly.

Example:

1. Start with the default diagram and fuse the two circled spiders
<img width="351" alt="Screenshot 2023-11-28 at 9 10 23 pm" src="https://github.com/Quantomatic/zxlive/assets/44865292/8f392d8a-af9c-4e30-a2b9-e9b1c8661839">

to get

<img width="341" alt="Screenshot 2023-11-28 at 9 10 28 pm" src="https://github.com/Quantomatic/zxlive/assets/44865292/ffbddba0-f423-4d1d-a82d-c4b3b0997371">

2. Save the proof and open it from the file again to get the following incorrect diagram
<img width="335" alt="Screenshot 2023-11-28 at 9 10 47 pm" src="https://github.com/Quantomatic/zxlive/assets/44865292/664e0549-d20e-4db4-841f-e7cb82031020">

**Comment by dlyongemallo** [comment]
I think this issue is more general than proofs and applies to all graphs. Maybe an issue should be filed named something like "vertices are renumbered on save/open graph, leading to bugs", and *this* specific issue with incorrect proof saving/loading should be considered blocked by that.

As noted in https://github.com/Quantomatic/zxlive/pull/213#issuecomment-1859230170, the numbering of the nodes can change when a graph is saved and then re-opened. 

The numbering of the nodes can matter, which means that the graph which is re-opened is functionally not the same graph as the one which was saved. For example, the "fuse spiders" rule currently fuses the selected vertices into the node with the lowest numbering. When the app is first opened, the nodes are numbered with 0, 1, 2, and 3 as the input nodes, 4, 5, 6, 7 in the second row, and so on. The first image in the first comment in this issue shows the fusion of vertices 14 and 18 by this ordering, which should fuse both into 14 (node on left), as shown here:

![Screenshot from 2023-12-17 22-22-48](https://github.com/Quantomatic/zxlive/assets/9248532/6cc8022a-855e-4eb2-b94d-e83d9dc0fed4|width=350)

This differs from the image shown above, which has fused them into the node on the right. (Presumably, this output was generated after the initial demo graph had been saved, then re-opened.) Of course, since "only connectivity matters", these are really the same diagram. Still, it might be unexpected behaviour for a user to carry out an operation on a graph, and get *visually* different results based on how it was saved/opened.

Similarly, when I re-open the proof, this is what I get:

![Screenshot from 2023-12-17 22-29-57](https://github.com/Quantomatic/zxlive/assets/9248532/aa1df754-5fd2-488f-a50f-b9c1fbaa4679|width=350)

This is the wrong graph, but it is a *different* wrong graph than above! Clearly, there is a bug in applying the diffs for proofs, which might be because some piece of code assumes the numbering will be kept the same.

**Comment by dlyongemallo** [comment]
I've isolated the source of the bug. It's this commit: https://github.com/Quantomatic/zxlive/commit/4bb81078bffc0501d8027585132b9d2272c34d7e.

If this commit is reverted, the proofs are saved and loaded correctly.

**Comment by dlyongemallo** [comment]
Of course, since that commit is intended to fix something (an error produced by saving a phase spider with a parameter), reverting it will reintroduce that issue. So either `to_json`/`from_json` have to be changed to fix this issue, or the parametrised phase spider issue will have to be fixed in some other way.

**Comment by jvdwetering** [comment]
This should now be fixed with the save file format in PyZX, where all vertices keep their specified index.
